doh-server/server.go

```diff
@@ -57,19 +57,40 @@ type DNSRequest struct {
 }
 
 func NewServer(conf *config) (s *Server) {
+	timeout := time.Duration(conf.Timeout) * time.Second
 	s = &Server{
 		conf: conf,
 		udpClient: &dns.Client{
 			Net:     "udp",
 			UDPSize: dns.DefaultMsgSize,
-			Timeout: time.Duration(conf.Timeout) * time.Second,
+			Timeout: timeout,
 		},
 		tcpClient: &dns.Client{
 			Net:     "tcp",
-			Timeout: time.Duration(conf.Timeout) * time.Second,
+			Timeout: timeout,
 		},
 		servemux: http.NewServeMux(),
 	}
+	if conf.LocalAddr != "" {
+		udpLocalAddr, err := net.ResolveUDPAddr("udp", conf.LocalAddr)
+		if err == nil {
+			s.udpClient.Dialer = &net.Dialer{
+				Timeout:   timeout,
+				LocalAddr: udpLocalAddr,
+			}
+		} else {
+			log.Println(err)
```

Comment:
Do you think an error here (e.g. if `local_addr` is malformed or missing a port number) should be fatal or just logged without halting execution (falling back to the default behavior as if `local_addr` were not specified)?

Reply:
I think a fatal error is better, since this error happens during initialization stage and we don't need to recover from this error.

What about your idea?

By the way, it is better to write `if err != nil { A } else { B }` than `if err == nil { B } else { A }` in Go.

Reply:
Thanks. Which would be more idiomatic in Go?

* `if err != nil { log.Fatalln(err) }` inside this function, or
* `return s, err` and log the fatal in `main`?

Reply:
I went with the latter.